### DISCUSSION TITLE
onClick not firing

### DISCUSSION THREAD

**Discussion by phidah** [vercel/next.js#15226] [Help] [answered]
I have a standard page with a link that I want to use to toggle the showing of another component:

```javascript
import React from 'react'

export default class Page extends React.Component {
  render(props) {
    <a onClick={() => console.log("wow")}>Click here</a>
  }
}

```

When clicking the link nothing happens in the console. I've tried various ways of doing this.

I tried doing it in a page directly, in a subcomponent, etc but it doesn't work at all. If I `console.log` in the contructor it gets displayed in the server logs.

What do I have to do to be able to make proper client side rendered React things in Next?

**Comment by jamesmosier** [comment, marked as answer]
That worked for me, although you have a syntax error with your render method:

```jsx
import React from 'react'

export default class Page extends React.Component {
  render() {
    return <a onClick={() => console.log("wow")}>Click here</a>
  }
}
```

Do you have any other code you could share?

**Reply by abubakar2029** [reply]
"use client"
import React from 'react'
const page = () => {
  function myFun(){
    console.log('helo');
  }
  return (
<>
  <h1>hey</h1>
  <button onClick={myFun}>hello</button>
</>
  )
}
export default page
Onclick is not working here?

**Comment by phidah** [comment]
Honestly don't know what was wrong. Taking your code directly did not work.
I took all the files from the project and copied to a new project and it worked.

**Comment by ifeora-emeka** [comment]
I'm having same issue as well `<h4 className="title" onClick={() => alert('yo.!!')}>
                    <a href='#c'>{data.title} </a>
                </h4>`

**Comment by lojithv** [comment]
In my case, I had to rename the file name of the component in lowercase. After that, all click listeners worked. Ex: Footer.js (old file name) -> footer.js

**Reply by nikovacs** [reply]
I renamed my file NavBar.tsx -> navBar.tsx -> NavBar.tsx and it worked... lol

**Comment by LukNasc** [comment]
i'm having same issue in the component, when i try in the `index.ts` file on root of `src/pages` dir work but not in `src/components/Component.ts`

_/src/components/Component.ts_
![image](https://user-images.githubusercontent.com/26725196/236586602-f41f2c68-b11f-47dd-8aec-101f0e3ad552.png)
**this not work**

_src/pages/index.ts_
![image](https://user-images.githubusercontent.com/26725196/236586707-d4469dc6-703b-4adf-bc54-baec2134ce32.png)
**this work**

**Reply by GoncaloFig** [reply]
I have the same issue...

Do you know the solution?

**Reply by abubakar2029** [reply]
I'm not sure but I think it was solved by updating node version

**Comment by Lagyu** [comment]
I was having the same issue, finally found that I was mistakenly loading the server side env variable from client side with `@t3-oss/env-nextjs`.
Removing the mistake made the `onClick` work again.

**Comment by runoncedev** [comment]
Was having the same issue, found that it was because I was doing it in [`_document.js` and it's only rendered on the server](https://nextjs.org/docs/pages/building-your-application/routing/custom-document). Moved the button somewhere else and it started working:

> This file is only rendered on the server, so event handlers like onClick cannot be used in _document.

**Reply by Adeen1** [reply]
love you brother. I was totally exhausted and could not find the solution to this problem but then I came across this comment after wasting 5 hours

**Comment by lunoxx** [comment]
> I have a standard page with a link that I want to use to toggle the showing of another component:
> 
> ```js
> import React from 'react'
> 
> export default class Page extends React.Component {
>   render(props) {
>     <a onClick={() => console.log("wow")}>Click here</a>
>   }
> }
> ```
> 
> When clicking the link nothing happens in the console. I've tried various ways of doing this.
> 
> I tried doing it in a page directly, in a subcomponent, etc but it doesn't work at all. If I `console.log` in the contructor it gets displayed in the server logs.
> 
> What do I have to do to be able to make proper client side rendered React things in Next?

It`s 2023 and problem is alive. Any chance to get solved closely ?

UPDATE:
I had node v16.3. I was too stupid to check this before.
If anyone else has this problem, UPDATE NODE.

**Reply by LukNasc** [reply]
The problem only happens when trying to place an event in a file outside the "pages" directory.
Perhaps using the "app" directory will solve the problem.

**Reply by seinwave** [reply]
Updating node worked for me, too. 

I had node v16.1. No click handlers worked. (Or other event handlers -- `mouseEnter`, `mouseLeave`, etc). 

But [upgrading node](https://medium.com/macoclock/update-your-node-js-on-your-mac-in-2020-948948c1ffb2) to 18.17.1 (stable release as of 9.8.2023) did the trick.

**Reply by suhail135** [reply]
You just saved me !!!!!!!!!!

**Reply by kokanek** [reply]
Had the same issue and it got solved by upgrading node version from 16. to 18. 

Never in a 1000 years would I have thought the issue is due to Node! What is the logic behind this behavior?

**Reply by SunriseXuu** [reply]
I was using error handling pattern suggested by nextjs forms and mutations docs. However, the form action in the client side component didn't respond at all when I press the submit button, nor switching to onclick. Upgrading node to 18 worked for me, too. Thanks a lot I didn't spend too much time to get here.

**Reply by Mitchell-Conrad-20** [reply]
Same issue for me. It strikes me as odd that the issue does not occur purely in React, rather, only in NextJS. Upgrading to the latest version of NodeJS (node-v18.17.1-x64) solved the problem.

**Reply by nathan-osman** [reply]
The amount of frustration and hair-pulling that I've gone through... and all for a solution as simple as this. 😆

**Reply by wadinj** [reply]
Thanks 👍 I found your gem after 1 hour - upgrade to v18 works like a charm

**Reply by syond** [reply]
For me it was just because I was rendering the element at `_document.jsx`. I moved to `_app.jsx`, following the ["Layout Pattern" section at the Nextjs documentation](https://nextjs.org/docs/pages/building-your-application/routing/pages-and-layouts#layout-pattern), and works fine.

**Comment by Shivansh-yadav13** [comment]
I forgot to add `<html>`  `<body>`  `<main>` in the `layout.tsx`
Adding them fixed the issue.

**Reply by CHENNI-Nidhaleddine** [reply]
Thanks it works for me

**Reply by iqrairfan100** [reply]
thanks so much

**Comment by xlebenny** [comment]
any hydration problem also can make entire(in anywhere) onclick doesn't work

I used 4 hours :(
